File: MobileAppReactNativeHooks/src/business/state/Stock.ts

/** Stocks the user owns. */
export class Stock {
  readonly ticker: string;
  readonly howManyShares: number;
  readonly averagePrice: number;

  constructor(ticker: string,
              howManyShares: number,
              averagePrice: number) {
    this.ticker = ticker;
    this.howManyShares = howManyShares;
    this.averagePrice = averagePrice;
  }

  copyWith({
             ticker = this.ticker,
             howManyShares = this.howManyShares,
             averagePrice = this.averagePrice
           }: {
    ticker?: string,
    howManyShares?: number,
    averagePrice?: number
  } = {}): Stock {
    return new Stock(ticker, howManyShares, averagePrice);
  }

  get costBasis() {
    return this.howManyShares * this.averagePrice;
  }

  get averagePriceStr() {
    return `US$ ${this.averagePrice.toFixed(2)}`;
  }

  toString(): string {
    return `${this.ticker}: ${this.howManyShares} shares`;
  }
}

File: MobileAppFlutterCelest/celest/lib/my_src/models/stock.dart
import 'package:meta/meta.dart';

import 'utils/json.dart';

/// Stocks the user has.
@immutable
class Stock {
  final String ticker;
  final int howManyShares;
  final double averagePrice;

  Stock(
    this.ticker, {
    required this.howManyShares,
    required this.averagePrice,
  });

  /// A stock with no shares.
  Stock.noStocks(String ticker) : this(ticker, howManyShares: 0, averagePrice: 0);

  double get costBasis => howManyShares * averagePrice;

  String get averagePriceStr => 'US\$ ${averagePrice.toStringAsFixed(2)}';

  @override
  String toString() => '$howManyShares $ticker @$averagePrice';

  Object? toJsonPersistor() => serialize<Stock>(this);

  factory Stock.fromJsonPersistor(Object? value) => deserialize<Stock>(value);

  // TODO: MARCELO
  // Map<String, dynamic> toJson() => {
  //       'ticker': ticker,
  //       'howManyShares': howManyShares,
  //       'averagePrice': averagePrice,
  //     };
  //
  // static Stock fromJson(Json json) {
  //   return Stock(
  //     json.asString('ticker')!,
  //     howManyShares: json.asInt('howManyShares')!,
  //     averagePrice: json.asDouble('averagePrice')!,
  //   );
  // }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Stock &&
          runtimeType == other.runtimeType &&
          ticker == other.ticker &&
          howManyShares == other.howManyShares &&
          averagePrice == other.averagePrice;

  @override
  int get hashCode => ticker.hashCode ^ howManyShares.hashCode ^ averagePrice.hashCode;
}
